### PR TITLE
[cli] Added generic CLI hook [fixes #262]

### LINKED ISSUE #262: Improve Command Line Argument Handling
We should look how different extensions can contribute to CLI arguments.

### PR DESCRIPTION
**Pull request by svenefftinge**
Signed-off-by: Sven Efftinge <sven.efftinge@typefox.io>

### CODE REVIEW

File: dev-packages/cli/src/generator/backend-generator.ts [hunk 1]
```diff
@@ -38,16 +37,18 @@ function load(raw) {
...
     application.use(express.static(path.join(__dirname, '../../lib'), {
         index: 'index.html'
     }));
-    return application.start(port, host);
+    return application.start();
 }
 
-module.exports = (port, host) => Promise.resolve()${this.compileBackendModuleImports(backendModules)}
-    .then(() => start(port, host)).catch(reason => {
+Promise.resolve()${this.compileBackendModuleImports(backendModules)}
```
Comment (akosyakov) [inline dev-packages/cli/src/generator/backend-generator.ts]: `module.exports = `? otherwise the port won't be propagated to the electron renderer process

Comment (akosyakov) [reply dev-packages/cli/src/generator/backend-generator.ts]: also wdio tests pass args programmatically

**Comment by svenefftinge** [comment]
This change also provides a good help and moves the workspace directory from an option to the primary param.

**Comment by svenefftinge** [comment]
@akosyakov I noticed that `theia browser` forwards any arguments appropriately, but unfortunately consumes the `--help`. Can we avoid that?

**Comment by akosyakov** [comment]
@svenefftinge Yargs allows disabling help with: `yargs.help(false)`. We can disable it if a command is provided. There is already such check in the dynamic system PR:
https://github.com/theia-ide/theia/blob/b2b8c0028d416128b5d45f574ae9803b922a2d31/dev-packages/cli/src/theia.ts#L70-L78

File: packages/core/src/node/backend-application.ts [hunk 2]
```diff
@@ -17,6 +20,23 @@ export interface BackendApplicationContribution {
     onStart?(server: http.Server): void;
 }
 
+@injectable()
+export class BackendApplicationCliContribution implements CliContribution {
+
+    port: number;
+    hostname: string | undefined;
```
Comment (kittaakos) [inline packages/core/src/node/backend-application.ts]: Did you intentionally leave these implicit `public`?

Comment (svenefftinge) [reply packages/core/src/node/backend-application.ts]: Yes, those properties are being accessed from outside.

**Pull request merged**